sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java

```diff
@@ -132,180 +160,72 @@ public boolean isEndRequired() {
-        long scheduleDelayMillis,
-        int maxQueueSize,
-        int maxExportBatchSize,
-        int exporterTimeoutMillis) {
-      this.spanExporter = spanExporter;
-      this.scheduleDelayMillis = scheduleDelayMillis;
-      this.maxQueueSize = maxQueueSize;
-      this.halfMaxQueueSize = maxQueueSize >> 1;
-      this.maxExportBatchSize = maxExportBatchSize;
-      this.exporterTimeoutMillis = exporterTimeoutMillis;
-      this.spansList = new ArrayList<>(maxQueueSize);
+  // Timeout is best effort don't need to handle the future return value.
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private void exportBatch() {
+    List<ReadableSpan> batch = new ArrayList<>(Math.min(maxExportBatchSize, queue.size()));
+    if (queue.drainTo(batch, maxExportBatchSize) == 0) {
+      scheduleNextExport();
+      return;
     }
 
-    private void addSpan(ReadableSpan span) {
-      synchronized (monitor) {
-        if (spansList.size() == maxQueueSize) {
-          droppedSpans.add(1);
-          return;
-        }
-        // TODO: Record a gauge for referenced spans.
-        spansList.add(span);
-        // Notify the worker thread that at half of the queue is available. It will take
-        // time anyway for the thread to wake up.
-        if (spansList.size() >= halfMaxQueueSize) {
-          monitor.notifyAll();
-        }
-      }
+    List<SpanData> forExport = new ArrayList<>(batch.size());
+    for (ReadableSpan span : batch) {
+      forExport.add(span.toSpanData());
     }
 
-    @Override
-    public void run() {
-      while (!Thread.currentThread().isInterrupted()) {
-        // Copy all the batched spans in a separate list to release the monitor lock asap to
-        // avoid blocking the producer thread.
-        ArrayList<ReadableSpan> spansCopy;
-        synchronized (monitor) {
-          // If still maxExportBatchSize elements in the queue better to execute an extra
-          if (spansList.size() < maxExportBatchSize) {
-            do {
-              // In the case of a spurious wakeup we export only if we have at least one span in
-              // the batch. It is acceptable because batching is a best effort mechanism here.
-              try {
-                monitor.wait(scheduleDelayMillis);
-              } catch (InterruptedException ie) {
-                // Preserve the interruption status as per guidance and stop doing any work.
-                Thread.currentThread().interrupt();
-                return;
-              }
-            } while (spansList.isEmpty());
+    final CompletableResultCode result = spanExporter.export(forExport);
+    result.whenComplete(
+        new Runnable() {
+          @Override
+          public void run() {
+            if (!result.isSuccess()) {
+              logger.log(Level.FINE, "Exporter failed");
+            }
+            if (queue.size() >= maxExportBatchSize) {
+              exportBatch();
+            } else {
+              scheduleNextExport();
```

Comment:
I am concerned, that if any error situation prevents the runtime to reach this line for one export attempt, then the whole BSP will stop working. It seem that current design (schedule next export when this one finishes) tries to achieve "fixed delay". If we switch to "fixed rate" we can simplify it. I don't see why "fixed rate" is worse than "fixed delay".